File: cpp/src/benchmarks/vector.cpp
#include "vector.h"
#include <cassert>

namespace bch {
void benchmark_vector1()
{
    std::vector<Object> vec1(500000, Object{});
    auto                vec2 = vec1;
    assert(vec1.size() == vec2.size());
}
void benchmark_vector2()
{
    std::vector<Object> vec1;
    for (int i = 0; i != 500000; ++i) {
        vec1.push_back(Object{});
    }
    assert(vec1.size() == 500000);
}
void benchmark_vector3()
{
    std::vector<Object> vec1;
    vec1.reserve(500000);
    for (int i = 0; i != 500000; ++i) {
        vec1.emplace_back();
    }
    assert(vec1.size() == 500000);
}

} // namespace bch

File: rust/src/benchmarks/vector.rs
use std::vec::Vec;

#[derive(Clone)]
struct Object {
    _id: i32,
    _version: String,
    _partners: Vec::<String>
}

impl Default for Object {
    fn default() -> Object {
        Object {
            _id: 0,
            _version: "Some version".to_string(),
            _partners: vec![
                "A".to_string(),
                "BB".to_string(),
                "CCC".to_string(),
                "DDDD".to_string(),
                "ZZZZZZZZZZZZZZZZZZZZZZZZZ".to_string()
                ]
        }
    }
}

pub fn benchmark_vector1() -> ()
{
    let vec1 = vec![Object::default(); 500000];
    let vec2 = vec1.clone();
    assert!(vec1.len() == vec2.len());
}

pub fn benchmark_vector2() -> ()
{
    let mut vec1 = vec![];
    for _i in 0..500000 {
        vec1.push(Object::default());
    }
    assert!(vec1.len() == 500000);
}

pub fn benchmark_vector3() -> ()
{
    let mut vec1 = vec![];
    vec1.reserve(500000);
    for _i in 0..500000 {
        vec1.push(Object::default());
    }
    assert!(vec1.len() == 500000);
}
